README.md
# 彦兮的博客

## 专栏

[优秀技术文章分享](https://github.com/18888628835/Blog/issues/62)

[掘金全集](https://juejin.cn/user/3940246036424679/posts)

## CSS 高级技巧案例

1.[CSS 编码技巧](https://github.com/18888628835/Blogs/issues/63)

2.[背景和边框](https://github.com/18888628835/Blog/blob/main/CSS高级技巧案例/背景和边框.md)

3.[形状](https://github.com/18888628835/Blog/blob/main/CSS高级技巧案例/形状.md)

4.[视觉效果](https://github.com/18888628835/Blog/blob/main/CSS高级技巧案例/视觉效果.md)

5.[结构与布局](https://github.com/18888628835/Blog/blob/main/CSS高级技巧案例/结构与布局.md)

6.[用户体验](https://github.com/18888628835/Blog/blob/main/CSS高级技巧案例/用户体验.md)

7.[动画效果](https://github.com/18888628835/Blog/blob/main/CSS高级技巧案例/动画效果.md)

## 开发技巧

1.[代码优化小技巧](https://github.com/18888628835/Blogs/issues/9)

2.[原生实现上传文件预览](https://github.com/18888628835/Blog/issues/46)

3.[原生实现图片懒加载](https://github.com/18888628835/Blog/issues/48)

4.[前端根据后端返回实现下载或者预览的几种方式](https://github.com/18888628835/Blog/issues/52)

5.[如何实现 JWT？](https://github.com/18888628835/Blog/issues/55)

6.[封装一个请求工具函数](https://github.com/18888628835/Blog/issues/59)

## 前端可视化

1.[如何使用 Echarts 和 SVG 文件写大屏地图](https://github.com/18888628835/Blog/blob/main/前端可视化/如何使用Echarts和SVG文件写大屏地图.md)

2.[Canvas](https://github.com/18888628835/Blog/blob/main/前端可视化/Canvas.md)

## JavaScript 设计模式

1.[JavaScript 设计模式与开发实践读书笔记(基础篇)](<https://github.com/18888628835/Blog/blob/main/JavaScript设计模式/JavaScript设计模式与开发实践读书笔记(基础篇).md>)

2.[单例模式、策略模式、代理模式、发布订阅模式、命令模式、组合模式](https://github.com/18888628835/Blog/blob/main/JavaScript设计模式/单例模式、策略模式、代理模式、发布订阅模式、命令模式、组合模式.md)

3.[模版方法模式、享元模式、职责链模式、中介者模式、装饰者模式、状态模式、适配器模式](https://github.com/18888628835/Blog/blob/main/JavaScript设计模式/模版方法模式、享元模式、职责链模式、中介者模式、装饰者模式、状态模式、适配器模式.md)

4.[设计原则和编程技巧](https://github.com/18888628835/Blog/blob/main/JavaScript设计模式/设计原则和编程技巧.md)

5.[代码重构](https://github.com/18888628835/Blog/blob/main/JavaScript设计模式/代码重构.md)

6.[JavaScript 设计模式小册](https://github.com/18888628835/Blog/blob/main/JavaScript设计模式/JavaScript设计模式小册.md)

## JavaScript

1.[let、const 与 var 的区别](https://github.com/18888628835/Blog/issues/11)

2.[解构赋值](https://github.com/18888628835/Blog/issues/12)

3.[扩展运算符](https://github.com/18888628835/Blog/issues/13)

4.[Set 和 Map 数据结构](https://github.com/18888628835/Blog/issues/18)

5.[Symbol](https://github.com/18888628835/Blog/issues/19)

6.[async 和 await](https://github.com/18888628835/Blog/issues/20)

7.[Iterator（迭代器)](https://github.com/18888628835/Blog/issues/38)

8.[对比 class 与旧构造函数的区别](https://github.com/18888628835/Blog/issues/39)

9.[模块(Module)](<https://github.com/18888628835/Blog/blob/main/JavaScript/模块(Module).md>)

10.[JSON](https://github.com/18888628835/Blog/blob/main/JavaScript/JSON.md)

11.[Generator，高级 Iteration](https://github.com/18888628835/Blog/blob/main/JavaScript/Generator.md)

12.[二进制数据和文件](https://github.com/18888628835/Blog/blob/main/JavaScript/二进制数据和文件.md)

13.[WebSocket](https://github.com/18888628835/Blog/blob/main/JavaScript/WebSocket.md#websocket简介)

14.[Fetch](https://github.com/18888628835/Blog/blob/main/JavaScript/Fetch.md)

## JavaScript 深入

1.[深入理解原型](https://github.com/18888628835/Blogs/issues/1)

2.[深入理解词法作用域](https://github.com/18888628835/Blogs/issues/2)

3.[深入理解函数的执行上下文栈](https://github.com/18888628835/Blogs/issues/3)

4.[深入理解变量对象](https://github.com/18888628835/Blogs/issues/4)

5.[深入理解作用域链](https://github.com/18888628835/Blogs/issues/5)

6.[深入理解执行上下文](https://github.com/18888628835/Blogs/issues/6)

7.[深入理解闭包](https://github.com/18888628835/Blogs/issues/7)

8.[深入理解垃圾回收](https://github.com/18888628835/Blogs/issues/8)

9.[深入浅出正则表达式](https://github.com/18888628835/Blog/issues/21)

10.[手动实现 new 关键字](https://github.com/18888628835/Blog/issues/40)

11.[手写 call、apply、bind](https://github.com/18888628835/Blog/issues/41)

12.[手写防抖和节流](https://github.com/18888628835/Blog/issues/42)

13.[函数 memory](https://github.com/18888628835/Blog/issues/43)

14.[函数柯里化](https://github.com/18888628835/Blog/issues/44)

15.[手写深拷贝](https://github.com/18888628835/Blog/issues/45)

## 浏览器

1.[浏览器加载文档和资源](https://github.com/18888628835/Blog/blob/main/浏览器/浏览器加载文档和资源.md)

2.[浏览器渲染过程](https://github.com/18888628835/Blog/blob/main/浏览器/浏览器渲染过程.md)

3.[Document](https://github.com/18888628835/Blog/blob/main/浏览器/Document.md)

4.[浏览器事件简介](https://github.com/18888628835/Blog/blob/main/浏览器/浏览器事件简介.md)

5.[UI 事件](https://github.com/18888628835/Blog/blob/main/浏览器/UI事件.md)

6.[事件循环](https://github.com/18888628835/Blog/blob/main/浏览器/事件循环.md)

7.[Frame 和 window](https://github.com/18888628835/Blog/blob/main/浏览器/Frame和window.md)

## Node.js

1.[MySQL 基础（DQL 单表查询）](https://github.com/18888628835/Blog/issues/58)

2.[MySQL 基础（DQL 多表查询）](https://github.com/18888628835/Blog/issues/60)

3.[MySQL 基础（高级概念和试验）](https://github.com/18888628835/Blog/issues/61)

4.[Sequelize 文档学习（基础篇）](https://github.com/18888628835/Blog/issues/56)

5.[Sequelize 文档学习（CRUD 查询篇）](https://github.com/18888628835/Blog/issues/57)

6.[Node.js 基础](https://github.com/18888628835/Blog/blob/main/Node.js/Node.js基础.md)

7.[GraphQL](https://github.com/18888628835/Blog/blob/main/Node.js/GraphQL.md)

## 前端拓展

1.[babel](https://github.com/18888628835/Blog/blob/main/前端拓展/babel.md)

2.[Webpack 从 0 开始配置开发环境](https://github.com/18888628835/Blog/blob/main/前端拓展/Webpack从0开始配置开发环境.md)

3.[前端性能优化](https://github.com/18888628835/Blog/blob/main/前端拓展/前端性能优化.md)

4.[前端工程化](https://github.com/18888628835/Blog/blob/main/前端拓展/前端工程化.md)

## 数据结构和算法系列

1.[复杂度分析](https://github.com/18888628835/Blogs/issues/10)

2.[线性表（数组）](https://github.com/18888628835/Blogs/issues/14)

3.[选择排序](https://github.com/18888628835/Blogs/issues/15)

4.[插入排序](https://github.com/18888628835/Blogs/issues/16)

5.[栈](https://github.com/18888628835/Blog/issues/17)

6.[队列](https://github.com/18888628835/Blog/issues/47)

7.[链表](https://github.com/18888628835/Blog/issues/49)

8.[理解递归](https://github.com/18888628835/Blog/issues/50)

9.[二叉搜索树](https://github.com/18888628835/Blog/issues/51)

10.[冒泡排序](https://github.com/18888628835/Blog/issues/53)

11.[快速排序](https://github.com/18888628835/Blog/issues/54)

12.[扁平化数组与树形结构的互转](https://github.com/18888628835/Blog/issues/65)

## 深入浅出 HTTP 协议

1.[HTTP 的由来](https://github.com/18888628835/Blog/issues/22)

2.[HTTP 是什么](https://github.com/18888628835/Blog/issues/23)

3.[HTTP 相关概念](https://github.com/18888628835/Blog/issues/24)

4.[TCP/IP 网络分层模型和 OSI 网络分层模型](https://github.com/18888628835/Blog/issues/25)

5.[域名系统](https://github.com/18888628835/Blog/issues/26)

6.[从 url 输入页面到页面展示到底发生了什么](https://github.com/18888628835/Blog/issues/27)

7.[HTTP 报文长什么样](https://github.com/18888628835/Blog/issues/28)

8.[请求方法](https://github.com/18888628835/Blog/issues/29)

9.[URI](https://github.com/18888628835/Blog/issues/30)

10.[HTTP 状态码](https://github.com/18888628835/Blog/issues/31)

11.[HTTP 的特点](https://github.com/18888628835/Blog/issues/32)

12.[Body 数据的内容协商](https://github.com/18888628835/Blog/issues/33)

13.[传输大文件的方法](https://github.com/18888628835/Blog/issues/34)

14.[Cookie 机制](https://github.com/18888628835/Blog/issues/35)

15.[HTTP 缓存控制](https://github.com/18888628835/Blog/blob/main/HTTP/HTTP缓存控制.md)

16.[HTTP 代理](https://github.com/18888628835/Blog/issues/37)

## Git

[Git](https://github.com/18888628835/Blog/blob/main/Git/Git.md)

## Framework

1.[vue3 composition-api](https://github.com/18888628835/Blog/blob/main/Framework/Vue3_basic.md)

2.[H5 手机 App 开发概念](https://github.com/18888628835/Blog/blob/main/Framework/H5手机App开发概念.md)

3.[Dart 基础语法](https://github.com/18888628835/Blog/blob/main/Framework/Dart基础语法.md)

4.[Flutter 渲染原理](https://github.com/18888628835/Blog/blob/main/Framework/Flutter渲染原理.md)

5.[Flutter 全局管理](https://github.com/18888628835/Blog/blob/main/Framework/Flutter全局管理.md)

## 勘误及提问

如果有错误或者不严谨的地方，请务必给予指正，十分感谢。

如果喜欢或者有所启发，欢迎 star，对作者也是一种鼓励。


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/18888628835/Blog/compare/v1.0.0...v1.1.0) (2022-05-23)


### Features

* add release-please docs ([a5b1b58](https://github.com/18888628835/Blog/commit/a5b1b58a4d82b466aab66a3efb42331ad95047ca))
* update release-please.yml ([3ebaa65](https://github.com/18888628835/Blog/commit/3ebaa6568d9ff5ec582e63cc37f2982a0b419af5))

## 1.0.0 (2022-05-23)


### Bug Fixes

* update subject status to required ([8179e55](https://github.com/18888628835/Blog/commit/8179e555f2970318be6a37a5393e2ae97a60b85b))


data-structure/树.ts
class TreeNode {
  key;
  left;
  right;
  constructor(key) {
    this.key = key;
    this.left = null;
    this.right = null;
  }
}
// 实现一个二叉搜索树
class BinarySearchTree {
  root;
  constructor() {
    this.root = null;
  }
  insertNode(root, key) {
    if (root.key > key) {
      if (root.left === null) {
        root.left = new TreeNode(key);
      } else {
        this.insertNode(root.left, key);
      }
    } else {
      if (root.right === null) {
        root.right = new TreeNode(key);
      } else {
        this.insertNode(root.right, key);
      }
    }
  }
  //插入节点
  insert(key) {
    if (this.root === null) {
      this.root = new TreeNode(key);
    } else {
      this.insertNode(this.root, key);
    }
  }
  //中序遍历的辅助函数
  private inOrderTraverseNode(node, callback) {
    if (node !== null) {
      this.inOrderTraverseNode(node.left, callback);
      callback(node.key);
      this.inOrderTraverseNode(node.right, callback);
    }
  }
  //中序遍历
  inOrderTraverse(callback) {
    this.inOrderTraverseNode(this.root, callback);
  }
  //先序遍历的辅助函数
  private preOrderTraverseNode(node, callback) {
    if (node !== null) {
      callback(node.key);
      this.preOrderTraverseNode(node.left, callback);
      this.preOrderTraverseNode(node.right, callback);
    }
  }
  // 先序遍历
  preOrderTraverse(callback) {
    this.preOrderTraverseNode(this.root, callback);
  }
  //后序遍历的辅助函数
  private postOrderTraverseNode(node, callback) {
    if (node !== null) {
      this.postOrderTraverseNode(node.left, callback);
      this.postOrderTraverseNode(node.right, callback);
      callback(node.key);
    }
  }
  //后序遍历
  postOrderTraverse(callback) {
    this.postOrderTraverseNode(this.root, callback);
  }
  minNode(node) {
    let current = node;
    while (current !== null && current.left !== null) {
      current = current.left;
    }
    return current;
  }
  min() {
    return this.minNode(this.root);
  }
  maxNode(node) {
    let current = node;
    while (current && current.right) {
      current = current.right;
    }
    return current;
  }
  max() {
    return this.maxNode(this.root);
  }
  searchNode(node, key) {
    if (node === null) {
      return false;
    }
    if (node.key === key) {
      return true;
    }
    if (key < node.key) {
      return this.searchNode(node.left, key);
    } else if (key > node.key) {
      return this.searchNode(node.right, key);
    }
  }
  search(key) {
    return this.searchNode(this.root, key);
  }
}
let tree = new BinarySearchTree();
tree.insert(11);
tree.insert(7);
tree.insert(15);
tree.insert(5);
tree.insert(3);
tree.insert(9);
tree.insert(8);
tree.insert(10);
tree.insert(13);
tree.insert(12);
tree.insert(14);
tree.insert(20);
tree.insert(18);
tree.insert(25);
tree.insert(6);
console.log(tree.search(26));
